File: AspNetMvcUnleashed/0672329980_source/Chapter19/CS/UnleashedBlog/UnleashedBlog.Tests/Controllers/ArchiveControllerTests.cs
using System;
using System.Collections.Generic;
using System.Web.Mvc;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using UnleashedBlog.Controllers;
using UnleashedBlog.Models;
using UnleashedBlog.Tests.Models;
using UnleashedBlog.Tests.Factories;

namespace UnleashedBlog.Tests.Controllers
{
    [TestClass]
    public class ArchiveControllerTests
    {
        private ControllerFactory _controllerFactory;
        private ArchiveController _archiveController;

        [TestInitialize]
        public void Initialize()
        {
            _controllerFactory = new ControllerFactory();
            _archiveController = _controllerFactory.GetArchiveController();
        }


        [TestMethod]
        public void IndexReturnsBlogEntriesByYear()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 11, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2011, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(2010, null, null, null);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }


        [TestMethod]
        public void IndexReturnsBlogEntriesByMonth()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 11, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(2010, 12, null, null);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }

        [TestMethod]
        public void IndexReturnsBlogEntriesByDay()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();

            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 25) ));
            bController.Create(BlogEntryFactory.GetWithDatePublished(new DateTime(2010, 12, 26) ));

            // Act
            var result = (ViewResult)_archiveController.Index(2010, 12, 25, null);

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(2, blogEntries.Count);
        }

        [TestMethod]
        public void IndexReturnsBlogEntryByName()
        {
            // Arrange
            var bController = _controllerFactory.GetBlogController();
            bController.Create(BlogEntryFactory.GetWithName("Test 1"));
            bController.Create(BlogEntryFactory.GetWithName("Test 2"));

            // Act
            var result = (ViewResult)_archiveController.Index(null, null, null, "Test-2");

            // Assert
            var blogEntries = (IList<BlogEntry>)result.ViewData.Model;
            Assert.AreEqual(1, blogEntries.Count);
            Assert.AreEqual("Test-2", blogEntries[0].Name);
        }


    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter20/VB/UnleashedBlog/UnleashedBlog.Tests/Controllers/ArchiveControllerTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports System.Web.Mvc

<TestClass()> _
Public Class ArchiveControllerTests

    Private _controllerFactory As ControllerFactory
    Private _archiveController As ArchiveController

    <TestInitialize()> _
    Public Sub Initialize()
        _controllerFactory = New ControllerFactory()
        _archiveController = _controllerFactory.GetArchiveController()
    End Sub


    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByYear()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 11, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2011, 12, 26)))

        ' Act
        Dim result As ViewResult = _archiveController.Index(Nothing, 2010, Nothing, Nothing, Nothing)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub


    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByMonth()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 11, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 26)))

        ' Act
        Dim result As ViewResult = _archiveController.Index(Nothing, 2010, 12, Nothing, Nothing)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub

    <TestMethod()> _
    Public Sub IndexReturnsBlogEntriesByDay()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()

        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 25)))
        bController.Create(BlogEntryFactory.GetWithDatePublished(New DateTime(2010, 12, 26)))

        ' Act
        Dim result As ViewResult = _archiveController.Index(Nothing, 2010, 12, 25, Nothing)

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(2, blogEntries.Count)
    End Sub

    <TestMethod()> _
    Public Sub IndexReturnsBlogEntryByName()
        ' Arrange
        Dim bController = _controllerFactory.GetBlogController()
        bController.Create(BlogEntryFactory.GetWithName("Test-1"))
        bController.Create(BlogEntryFactory.GetWithName("Test-2"))

        ' Act
        Dim result As ViewResult = _archiveController.Index(Nothing, Nothing, Nothing, Nothing, "Test-2")

        ' Assert
        Dim blogEntries = CType(result.ViewData.Model, IList(Of BlogEntry))
        Assert.AreEqual(1, blogEntries.Count)
        Assert.AreEqual("Test-2", blogEntries(0).Name)
    End Sub


    ' Ajax tests

    <TestMethod()> _
    Public Sub Index_AjaxReturnsPartialViewResult()
        ' Act
        Dim result = _archiveController.Index_Ajax(0, 2010, 12, 25, "TheName")

        ' Assert
        Assert.IsInstanceOfType(result, GetType(PartialViewResult))
    End Sub


End Class
